plugins/woocommerce/src/Internal/Utilities/PluginInstaller.php

```diff
@@ -0,0 +1,317 @@
+					$install_ok = false;
+				} else {
+					$plugin_version = $plugin_info['Version'];
+				}
+			}
+		}
+
+		$installed_by = $metadata['installed_by'] ?? 'WooCommerce';
+		if ( $install_ok ) {
+			$result['plugin_name'] = $plugin_name;
+			$plugin_data           = array(
+				'version' => $plugin_version,
+				'date'    => current_time( 'mysql' ),
+			);
+			if ( ! empty( $metadata ) ) {
+				$plugin_data['metadata'] = $metadata;
+			}
+
+			$post_install = function () use ( $plugin_name, $plugin_version, $installed_by, $plugin_url, $plugin_data ) {
+				$auto_installed_plugins                 = get_option( 'woocommerce_autoinstalled_plugins', array() );
+				$auto_installed_plugins[ $plugin_name ] = $plugin_data;
+				$log_context                            = array(
+					'source'        => 'plugin_auto_installs',
+					'recorded_data' => $plugin_data,
+				);
+				update_option( 'woocommerce_autoinstalled_plugins', $auto_installed_plugins );
+				wc_get_logger()->info( "Plugin $plugin_name v{$plugin_version} installed by $installed_by, source: $plugin_url", $log_context );
+			};
+		} else {
+			$messages     = $skin->get_upgrade_messages();
+			$post_install = function () use ( $plugin_name, $plugin_url, $installed_by, $messages ) {
+				$log_context   = array(
+					'source'             => 'plugin_auto_installs',
+					'installer_messages' => $messages,
+				);
+				$plugin_name ??= '';
+				wc_get_logger()->error( "$installed_by failed to install plugin {$plugin_name} from source: $plugin_url", $log_context );
+			};
+		}
+
+		$this->run_callback_in_all_sites( $post_install );
+
+		$result['install_ok'] = $install_ok ?? false;
+		return $result;
+	}
+
+	/**
+	 * Get the information provided by 'get_plugins' for a given plugin.
+	 * The returned array will also contain an extra boolean key, '_is_active_.
+	 *
+	 * @param string $plugin_name The name of the plugin to get information for.
+	 * @return array|null An array with plugin information, or null if no plugin is installed with that name.
+	 */
+	private function get_installed_plugin_info( string $plugin_name ): ?array {
+		wp_clean_plugins_cache();
+
+		$plugin_name = plugin_basename( $plugin_name );
+		$all_plugins = get_plugins();
+
+		$plugin_info = $all_plugins[ $plugin_name ] ?? null;
+		if ( is_null( $plugin_info ) ) {
+			return null;
+		}
+
+		$plugin_info['_is_active'] = in_array( $plugin_name, array_map( 'plugin_basename', wp_get_active_and_valid_plugins() ), true );
+		return $plugin_info;
+	}
+
+	/**
+	 * Run a callback in each existing site (if multisite) or just once (if single site).
+	 *
+	 * @param callable $callback The callback to run.
+	 */
+	private static function run_callback_in_all_sites( callable $callback ) {
+		if ( ! is_multisite() ) {
+			$callback();
+			return;
+		}
+
+		foreach ( get_sites() as $site ) {
+			switch_to_blog( $site->blog_id );
+			try {
+				$callback();
+			} finally {
+				restore_current_blog();
+			}
+		}
```

Comment:
Not sure this will perform well on big networks. I understand this is done to set the option and also log that the plugin was installed. I'm not entirely sure about the latter, but for the former, couldn't we use `get_site_option()` in the MU context instead of saving an option to each site in the network?

Reply:
We need to do the blog switch for the logging anyway, and it doesn't look like `get_site_option` would provide a big performance gain compared to `switch`/`restore` as it's doing a lot of processing too: https://developer.wordpress.org/reference/functions/get_network_option/